dagstore_control.go

```diff
@@ -341,27 +344,25 @@ func (d *DAGStore) control() {
 
 		s.lk.Unlock()
 	}
-
-	if err != context.Canceled {
-		log.Errorw("consuming next task failed; aborted event loop; dagstore unoperational", "error", err)
-	}
 }
 
-func (d *DAGStore) consumeNext() (tsk *task, error error) {
+func (d *DAGStore) consumeNext() (tsk *task, gc chan *GCResult, error error) {
 	select {
 	case tsk = <-d.internalCh: // drain internal first; these are tasks emitted from the event loop.
-		return tsk, nil
+		return tsk, nil, nil
 	case <-d.ctx.Done():
-		return nil, d.ctx.Err() // TODO drain and process before returning?
+		return nil, nil, d.ctx.Err() // TODO drain and process before returning?
 	default:
 	}
 
 	select {
 	case tsk = <-d.externalCh:
-		return tsk, nil
+		return tsk, nil, nil
 	case tsk = <-d.completionCh:
-		return tsk, nil
+		return tsk, nil, nil
+	case gc := <-d.gcCh:
```

Comment:
This can race with the shard state book-keeping/reconciliation we do on startup because we submit ops to the external channel on startup.  Since the ordering inside a select is indeterminate, we could end up picking the gc channel here (if client calls GC after startup) over the requests queued in the external channel at startup.

I think it would be nice to maintain the property of finishing all ops the dagstore itself submits to the event loop on startup before we pick client requests to avoid shooting ourselves in the foot down the line.

Reply:
> I think it would be nice to maintain the property of finishing all ops the dagstore itself submits to the event loop on startup before we pick client requests to avoid shooting ourselves in the foot down the line.

We do that, by consuming from internalCh first.


> This can race with the shard state book-keeping/reconciliation we do on startup because we submit ops to the external channel on startup. Since the ordering inside a select is indeterminate, we could end up picking the gc channel here (if client calls GC after startup) over the requests queued in the external channel at startup.

I don't think that's wrong. In fact, I was considering the opposite: prioritising GC over external operations. The user may be executing GC because the system is under pressure, and if not run on time, other operations may fail.

Reply:
@raulk 

Yes, but the problem is that the startup operations are being submitted to the external channel.

```
// release the queued registrations before we return.
	for _, s := range toRegister {
		_ = dagst.queueTask(&task{op: OpShardRegister, shard: s, waiter: &waiter{ctx: ctx}}, dagst.externalCh)
	}

	// queue shard recovery for shards in the errored state before we return.
	for _, s := range toRecover {
		_ = dagst.queueTask(&task{op: OpShardRecover, shard: s, waiter: &waiter{ctx: ctx}}, dagst.externalCh)
	}
```

Now that I think of it, these look like they should be submitted to the internal channel.

Reply:
GC can't be called at that point because we haven't even returned the instance to the user.

Reply:
The only thing that could happen is that operations remained queued when we return the DAGStore, and the user immediately calls GC. However, I'm not sure how that would be offensive.

Reply:
What I'm saying is that even after this request is queued,  if a user submits a GC request, it's possible for the GC request to be picked over the startup ops. This will break the property of "startup ops before all client calls".

Reply:
@raulk It's probably not offensive today. But, I'd like for us to maintain the invariant of all startup operations being executed before any clients ops are executed.

Reply:
That's true for GC or non-GC operations. That property is not formalised. If we want to introduce that property (and I think it's a good idea), we should pass a chan ShardResult in the waiters, and wait until we've consumed all responses. However, we might not want to block the constructor if we do that (synchronous recovery can take long). The implied contract of a constructor is that it just constructs the object, it does not do heavy lifting. Instead, we may want to move this logic to a `Start` method.

Reply:
How about simply submitting all startup ops to the internal channel ? That too should give us this property, right ?